streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/HistoryMetricsService.scala

```diff
@@ -0,0 +1,280 @@
+
+        queue.addFirst(HistoryMetricsItem(now, inputMetrics))
+
+        // remove old data
+        if (queue.size() > retainCount) {
+          queue.removeLast()
+        }
+      }
+    }
+
+    def read: List[HistoryMetricsItem] = {
+      val result = new ListBuffer[HistoryMetricsItem]
+      import scala.collection.JavaConversions.asScalaIterator
+      queue.iterator().foreach(result.prepend(_))
+      result.toList
+    }
+  }
+
+  // we will retain at max @RETAIN_HISTORY_HOURS history data
+  val RETAIN_HISTORY_DATA_HOURS = 72 // hours
+
+  // time interval between two history data points.
+  val RETAIN_HISTORY_DATA_INTERVAL_MS = 3600 * 1000 //ms
+
+  // we will retain at max @RETAIN_LATEST_SECONDS recent data points
+  val RETAIN_RECENT_DATA_SECONDS = 300 // s
+
+  // time interval between two recent data points.
+  val RETAIN_RECENT_DATA_INTERVAL_MS = 15 * 1000 //ms
+
+  class HistogramMetricsStore extends MetricsStore {
+
+    private val compartor = (left: HistoryMetricsItem, right: HistoryMetricsItem) =>
+      left.value.asInstanceOf[Histogram].mean > right.value.asInstanceOf[Histogram].mean
+
+    private val history = new MinMaxMetricsStore(
+      RETAIN_HISTORY_DATA_HOURS, RETAIN_HISTORY_DATA_INTERVAL_MS, compartor)
+
+    private val recent = new SingleValueMetricsStore(
+      RETAIN_RECENT_DATA_SECONDS * 1000 / RETAIN_RECENT_DATA_INTERVAL_MS, RETAIN_RECENT_DATA_INTERVAL_MS)
+
+    override def add(inputMetrics: MetricType): Unit = {
+      recent.add(inputMetrics)
+      history.add(inputMetrics)
+    }
+
+    override def read: List[HistoryMetricsItem] = {
+      history.read ++ recent.read
+    }
+  }
+
+  class MeterMetricsStore extends MetricsStore {
+
+    private val compartor = (left: HistoryMetricsItem, right: HistoryMetricsItem) =>
+      left.value.asInstanceOf[Meter].meanRate > right.value.asInstanceOf[Meter].meanRate
+
+    private val history = new MinMaxMetricsStore(
+      RETAIN_HISTORY_DATA_HOURS, RETAIN_HISTORY_DATA_INTERVAL_MS, compartor)
+
+    private val recent = new SingleValueMetricsStore(
+      RETAIN_RECENT_DATA_SECONDS * 1000 / RETAIN_RECENT_DATA_INTERVAL_MS, RETAIN_RECENT_DATA_INTERVAL_MS)
+
+    override def add(inputMetrics: MetricType): Unit = {
+      recent.add(inputMetrics)
+      history.add(inputMetrics)
+    }
+
+    override def read: List[HistoryMetricsItem] = {
+      history.read ++ recent.read
+    }
+  }
+
+  class CounterMetricsStore extends MetricsStore {
+
+    private val history = new SingleValueMetricsStore(RETAIN_HISTORY_DATA_HOURS, RETAIN_HISTORY_DATA_INTERVAL_MS)
+
+    private val recent = new SingleValueMetricsStore(
+      RETAIN_RECENT_DATA_SECONDS * 1000 / RETAIN_RECENT_DATA_INTERVAL_MS,
+      RETAIN_RECENT_DATA_INTERVAL_MS)
+
+    override def add(inputMetrics: MetricType): Unit = {
+      history.add(inputMetrics)
+      recent.add(inputMetrics)
+    }
+
+    override def read: List[HistoryMetricsItem] = {
+      history.read ++ recent.read
+    }
+  }
+}
```

Comment:
After rethinking this, I will make them configurable, and add a comment that this value should not be set to too big, due to implementation limit.